source/extensions/clusters/dynamic_forward_proxy/cluster.cc

```diff
@@ -161,11 +161,18 @@ Cluster::LoadBalancer::chooseHost(Upstream::LoadBalancerContext* context) {
     return nullptr;
   }
 
+  std::string sni;
   absl::string_view host;
   if (context->downstreamHeaders()) {
     host = context->downstreamHeaders()->getHostValue();
   } else if (context->downstreamConnection()) {
-    host = context->downstreamConnection()->requestedServerName();
+    // Embed downstream TCP connection destination port into SNI hostname. This is not necessary
+    // when using the Host header since the port is embeded by the client
+    // https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Host
+    sni = std::string(context->downstreamConnection()->requestedServerName()).c_str();
+    uint32_t port = context->downstreamConnection()->addressProvider().localAddress()->ip()->port();
+    absl::StrAppend(&sni, ":", port);
+    host = sni;
```

Comment:
You could use `host = absl::StrCat(...` to concatenate the requestedServerName, ":", and port directly.
(and below similarly, the c_str() thing is a

Reply:
I seem to get memory corruption whenever I use `StrCat`. For example in `onNewConnection()` I tried:

```
host = absl::StrCat(sni, ":",
      read_callbacks_->connection().addressProvider().localAddress()->ip()->port());
```
but for some longer domains the host variable becomes corrupted (for example `curl https://www.washingtonpost.com --resolve www.washingtonpost.com:443:127.0.0.1` with Envoy running locally). I don't really know why though.

Reply:
I think I have addressed your comment as you were imagining now.

I think I need some help fixing the tests though. Since the upstream connection destination TCP port is using the downstream connection destination port now, there isn't a way to hard code the destination port to 443 like the tests are currently (by setting the `port_value` field). However, the listener port seems to be dynamic (bind to 0) so I'm not sure the best way to handle that.